### PR TITLE
Fix a crash if ``str.format`` is referenced without being called

### LINKED ISSUE #5058: Crash on pandas data_frame 'to_string'
### Bug description

```python
When parsing the following file:

"""Definition of the base evaluation result class."""

from collections import defaultdict
from typing import AbstractSet, Dict, List, Optional, Union

import numpy as np
from pandas import DataFrame
from pydantic import BaseModel, PrivateAttr

Scores = Dict[str, Union[int, float]]
ScoresList = List[Scores]
AVERAGE = "AVERAGE"
OVERALL = "OVERALL"


class Result(BaseModel):
    """The base class for bdd100k evaluation results.

    Each data field corresponds to a evluation metric. The value for each
    metric is a list of dicts, each dict maps the category names to scores.
    There used to be two or three dicts in the list. The first one contains
    keys of basic categories, and the last one contains conclusion categories
    like 'OVERALL' and 'AVERAGE'. The middle one (optional), contains super
    classes for the two-level class hierarchy case.

    Functions:
        {} -> dict[str, dict[str, int | float]]:
            export all data to a nested dict.
        json() -> str:
            export the nested dict of `dict()` into a JSON string.
        pd_frame() -> pandas.DataFrame:
            export data fields to a formatted DataFrame.
        table() -> str:
            export data fields to a formatted table string.
        summary() -> dict[str, int | float]:
            export most important fields to a flattened dict.
        __str__() -> str:
            the same as `table()`.
    """

    _row_breaks: List[int] = PrivateAttr([])

    def __init__(self, **data: Union[int, float, ScoresList]) -> None:
        """Check the input structure and initiliaze the model.

        All keys in the Scoreslist need to be the same set for the different
        evaluation metrics.
        """
        data_check: Dict[str, ScoresList] = {
            metric: cont
            for metric, cont in data.items()
            if isinstance(cont, list)
        }
        ref_scores_list = data_check[list(data_check.keys())[0]]
        for scores_list in data_check.values():
            assert len(scores_list) == len(ref_scores_list)
            for scores, ref_scores in zip(scores_list, ref_scores_list):
                assert scores.keys() == ref_scores.keys()
        super().__init__(**data)
        cur_index = 1
        self._row_breaks = [1]
        for scores in ref_scores_list[:-1]:
            cur_index += 1 + len(scores)
            self._row_breaks.append(cur_index)

    def __eq__(self, other: "Result") -> bool:  # type: ignore
        """Check whether two instances are equal."""
        if self._row_breaks != other._row_breaks:
            return False
        other_dict = dict(other)
        for metric, scores_list in self:
            other_scores_list = other_dict[metric]
            if not isinstance(scores_list, list):
                if scores_list != other_scores_list:
                    return False
                continue
            if len(scores_list) != len(other_scores_list):
                return False
            for scores, other_scores in zip(scores_list, other_scores_list):
                if set(scores.keys()) != set(other_scores.keys()):
                    return False
                for category, score in scores.items():
                    if not np.isclose(score, other_scores[category]):
                        return False
        return super().__eq__(other)

    def pd_frame(
        self,
        include: Optional[AbstractSet[str]] = None,
        exclude: Optional[AbstractSet[str]] = None,
    ) -> DataFrame:
        """Convert data model into a data frame.

        Args:
            include (set[str]): Optional, the metrics to convert
            exclude (set[str]): Optional, the metrics not to convert
        Returns:
            data_frame (pandas.DataFrmae): the exported DataFrame
        """
        frame_dict: Dict[str, Scores] = defaultdict(dict)
        for metric, scores_list in self.dict(
            include=include, exclude=exclude  # type: ignore
        ).items():
            if not isinstance(scores_list, list):
                continue
            for scores in scores_list:
                for cls_, score in scores.items():
                    frame_dict[metric][cls_] = score
        return DataFrame.from_dict(frame_dict)

    def table(
        self,
        include: Optional[AbstractSet[str]] = None,
        exclude: Optional[AbstractSet[str]] = None,
    ) -> str:
        """Convert data model into a table for formatted printing.

        Args:
            include (set[str]): Optional, the metrics to convert
            exclude (set[str]): Optional, the metrics not to convert
        Returns:
            table (str): the exported table string
        """
        data_frame = self.pd_frame(include, exclude)
        summary = data_frame.to_string(float_format="{:.1f}".format)
        summary = summary.replace("NaN", " - ")
        strs = summary.split("\n")
        split_line = "-" * len(strs[0])

        for row_ind in self._row_breaks:
            strs.insert(row_ind, split_line)
        summary = "".join([f"{s}\n" for s in strs])
        summary = "\n" + summary
        return summary  # type: ignore

    def __str__(self) -> str:
        """Convert the data into a printable string."""
        print("hi{}".format("world"))
        return self.table()

    def summary(
        self,
        include: Optional[AbstractSet[str]] = None,
        exclude: Optional[AbstractSet[str]] = None,
    ) -> Scores:
        """Convert the data into a flattened dict as the summary.

        This function is different to the `.dict()` function.
        As a comparison, `.dict()` will export all data fields as a nested
        dict, While `.summary()` only exports most important information,
        like the overall scores, as a flattened compact dict.

        Args:
            include (set[str]): Optional, the metrics to convert
            exclude (set[str]): Optional, the metrics not to convert
        Returns:
            dict[str, int | float]: returned summary of the result
        """
        summary_dict: Dict[str, Union[int, float]] = {}
        for metric, scores_list in self.dict(
            include=include, exclude=exclude  # type: ignore
        ).items():
            if not isinstance(scores_list, list):
                summary_dict[metric] = scores_list
            else:
                summary_dict[metric] = scores_list[-1].get(
                    OVERALL, scores_list[-1].get(AVERAGE)
                )
        return summary_dict
```

pylint crashed with a ``AttributeError`` and with the following stacktrace:
```
Traceback (most recent call last):
  File "/itet-stor/tobfischer/net_scratch/conda_envs/openmt/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1008, in _check_files
    self._check_file(get_ast, check_astroid_module, file)
  File "/itet-stor/tobfischer/net_scratch/conda_envs/openmt/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1043, in _check_file
    check_astroid_module(ast_node)
  File "/itet-stor/tobfischer/net_scratch/conda_envs/openmt/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1180, in check_astroid_module
    retval = self._check_astroid_module(
  File "/itet-stor/tobfischer/net_scratch/conda_envs/openmt/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1227, in _check_astroid_module
    walker.walk(node)
  File "/itet-stor/tobfischer/net_scratch/conda_envs/openmt/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 78, in walk
    self.walk(child)
  File "/itet-stor/tobfischer/net_scratch/conda_envs/openmt/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 78, in walk
    self.walk(child)
  File "/itet-stor/tobfischer/net_scratch/conda_envs/openmt/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 78, in walk
    self.walk(child)
  [Previous line repeated 4 more times]
  File "/itet-stor/tobfischer/net_scratch/conda_envs/openmt/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 75, in walk
    callback(astroid)
  File "/itet-stor/tobfischer/net_scratch/conda_envs/openmt/lib/python3.8/site-packages/pylint/checkers/refactoring/recommendation_checker.py", line 327, in visit_const
    self._detect_replacable_format_call(node)
  File "/itet-stor/tobfischer/net_scratch/conda_envs/openmt/lib/python3.8/site-packages/pylint/checkers/refactoring/recommendation_checker.py", line 340, in _detect_replacable_format_call
    if node.parent.parent.args:
AttributeError: 'Keyword' object has no attribute 'args'
```

NOTE: When i comment the line summary = data_frame.to_string(float_format="{:.1f}".format), the error disappears

### Configuration

```ini
https://github.com/scalabel/scalabel/blob/master/.pylintrc
```

### Command used

```shell
pylint scalabel
```

### Pylint output

```shell
Exception on node <Const.str l.125 at 0x7fd2dcae8520> in file '/home/tobfischer/PycharmProjects/scalabel/scalabel/eval/result.py'
Traceback (most recent call last):
  File "/itet-stor/tobfischer/net_scratch/conda_envs/openmt/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 75, in walk
    callback(astroid)
  File "/itet-stor/tobfischer/net_scratch/conda_envs/openmt/lib/python3.8/site-packages/pylint/checkers/refactoring/recommendation_checker.py", line 327, in visit_const
    self._detect_replacable_format_call(node)
  File "/itet-stor/tobfischer/net_scratch/conda_envs/openmt/lib/python3.8/site-packages/pylint/checkers/refactoring/recommendation_checker.py", line 340, in _detect_replacable_format_call
    if node.parent.parent.args:
AttributeError: 'Keyword' object has no attribute 'args'
************* Module scalabel.eval.result
scalabel/eval/result.py:1:0: F0001: Fatal error while checking 'scalabel/eval/result.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in '/scratch/.cache_tobfischer/pylint/pylint-crash-2021-09-21-13.txt'. (fatal)
```

### Expected behavior

should pass

### Pylint version

```shell
2.11
```

### OS / Environment

_No response_

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by timmartin**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

If the `str.format` function was being referenced (not called) in the keyword argument to a function, the `consider-using-f-string` checker was crashing due to attempting to obtain the `.args` from something that wasn't a call object. Referencing format without calling it in this way might be a mistake, but that should probably be another checker.

Closes #5058

### CODE REVIEW

File: pylint/checkers/refactoring/recommendation_checker.py [hunk 1]
```diff
@@ -345,6 +345,10 @@ def _detect_replacable_format_call(self, node: nodes.Const) -> None:
             if isinstance(node.parent.parent, nodes.Assign):
                 return
 
+            # If the format function is not called, ignore this
+            if not isinstance(node.parent.parent, nodes.Call):
```
Comment (DanielNoord) [inline pylint/checkers/refactoring/recommendation_checker.py:349]: I'm wondering if we can't merge this with the check on L345. If the check on L349 passes, L345 should pass as well. We would only need to add both use cases to the comment on L344.

File: tests/functional/c/consider/consider_using_f_string.py [hunk 2]
```diff
@@ -40,6 +40,11 @@ def print_good():
     print("%s" % "\n")
     print("%s" % "\n".join(i for i in "string"))
 
+    def wrap_print(value):
```
Comment (DanielNoord) [inline tests/functional/c/consider/consider_using_f_string.py:43]: I think we can move this to the bottom of this file. Perhaps with a comment saying:
`Non valid code that shouldn't crash the checker`.
That would decrease the size of this diff as all line numbers don't change.

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Thanks for this fix and your investigation in #5058. I initially made this checker and am really happy you found this!

I requested two small changes which I think would improve this commit/merge, but the general idea of the commit is good!

**Review by Pierre-Sassoulas — APPROVED** [review]
Great analysis and fix @timmartin , thank you. I approve with @DanielNoord's suggestion especially the one that will reduce the diff.

**Comment by DanielNoord** [comment]
You'll need to merge main to fix the merge conflict, otherwise this looks good to go!

File: ChangeLog [hunk 3]
```diff
@@ -58,6 +58,10 @@ Release date: TBA
 
   Closes #3614
 
+* Fixed crash in ``consider-using-f-string`` if ``format`` is not called
```
[the remark this answers is not in the document]
Comment (Pierre-Sassoulas) [reply ChangeLog:61]: Imo order do not matter at all. I even hesitated at some point to explicitly tell contributors to input them at random location so as to reduce merge conflict when another changelog entry is merged in main.

**Pull request merged**